### ISSUE TITLE
vcpkg: too long file path for gcc?

### ISSUE THREAD

**Issue by froha** [microsoft/vscode-embedded-tools#37]
Type: Bug Report

**Describe the bug**
- OS and Version: Windows 10 21H2
- VS Code Version: 1.75.1
- Embedded Tools Extension Version: 0.6.0

**To Reproduce**
Download your example project:
[https://github.com/azure-rtos/getting-started](https://github.com/azure-rtos/getting-started)
Open following folder:
STMicroelectronics/B-L4S5I-IOT01A
Wait for the integrated vcpkg to activate the project.
Build

**Expected behavior**
No errors

**Code sample and logs**
`[build] [1/1294] Building C object lib/threadx/CMakeFiles/threadx.dir/common/src/tx_thread_info_get.c.obj
[build] FAILED: lib/threadx/CMakeFiles/threadx.dir/common/src/tx_thread_info_get.c.obj 
[build] C:\Users\Abcdefghijk.Lmnopqrs\AppData\Roaming\Code\User\globalStorage\ms-vscode.vscode-embedded-tools\vcpkg\root\downloads\artifacts\vcpkg-ce-default\compilers.arm.gcc\2020.10.0\bin\arm-none-eabi-gcc.exe -DTX_INCLUDE_USER_DEFINE_FILE -IC:/code/getting-started/shared/lib/threadx/ports/cortex_m4/gnu/inc -Ilib/threadx/custom_inc -isystem C:/code/getting-started/shared/lib/threadx/common/inc -mthumb -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -g3 -ffunction-sections -fdata-sections -fno-strict-aliasing -fno-builtin -fno-common -Wall -Wshadow -Wdouble-promotion -Werror -Wno-unused-parameter -O0 -std=gnu99 -MD -MT lib/threadx/CMakeFiles/threadx.dir/common/src/tx_thread_info_get.c.obj -MF lib\threadx\CMakeFiles\threadx.dir\common\src\tx_thread_info_get.c.obj.d -o lib/threadx/CMakeFiles/threadx.dir/common/src/tx_thread_info_get.c.obj -c C:/code/getting-started/shared/lib/threadx/common/src/tx_thread_info_get.c
[build] In file included from C:/code/getting-started/shared/lib/threadx/ports/cortex_m4/gnu/inc/tx_port.h:85,
[build]                  from C:/code/getting-started/shared/lib/threadx/common/inc/tx_api.h:121,
[build]                  from C:/code/getting-started/shared/lib/threadx/common/src/tx_thread_info_get.c:28:
[build] c:\users\Abcdefghijk.Lmnopqrs\appdata\roaming\code\user\globalstorage\ms-vscode.vscode-embedded-tools\vcpkg\root\downloads\artifacts\vcpkg-ce-default\compilers.arm.gcc\2020.10.0\arm-none-eabi\include\stdlib.h:10:10: fatal error: machine/ieeefp.h: No such file or directory
[build]    10 | #include <machine/ieeefp.h>
[build]       |          ^~~~~~~~~~~~~~~~~~
[build] compilation terminated.`

**Additional context**
Too long file path?
`C:\Users\Abcdefghijk.Lmnopqrs\AppData\Roaming\Code\User\globalStorage\ms-vscode.vscode-embedded-tools\vcpkg\root\downloads\artifacts\vcpkg-ce-default\compilers.arm.gcc\2020.10.0\arm-none-eabi\include\machine\ieeefp.h`
Already mention it in the issue microsoft/vcpkg#29815 , but it should be a new issue.
Possible solution: enable to move the vcpkg root directory

**Comment by benmcmorran** [comment]
Thanks for the report! To help narrow down if long paths are causing the issue, can you try [enabling long path support](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=powershell#enable-long-paths-in-windows-10-version-1607-and-later) (if it's not already enabled) and then restarting?

```powershell
New-ItemProperty -Path "HKLM:\SYSTEM\CurrentControlSet\Control\FileSystem" -Name "LongPathsEnabled" -Value 1 -PropertyType DWORD -Force
```

I'm also working on a change to enable overriding the storage location for the vcpkg installation, which should be available on the pre-release channel within a few days.

**Comment by froha** [comment]
Long path support was enbaled.
Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled = 1

A colleague has the same error, but the long path support was not enabled.
Enabling it and restart his pc didn't change anything.

**Comment by froha** [comment]
Additinal info:
a small test with vcpkg

Working:
$Env:VCPKG_ROOT = "C:\folder0asd\folder1asdfasdf\folder2\folder3\fol4\fol5\folder6asdfas\folder7asdfasdfasdfasdfasdfas"

Not working:
$Env:VCPKG_ROOT = "C:\folder0asd\folder1asdfasdf\folder2\folder3\fol4\fol5\folder6asdfas\folder7asdfasdfasdfasdfasdfasd"

Didn't find anything about a limit for gcc, but i had not much time today.

**Comment by benmcmorran** [comment]
Your most recent comment looks like this is definitely a long path issue. I'm hoping to have a change in the pre-release channel tomorrow that enables overriding the vcpkg storage location. I'll let you know once it's available for testing, and thanks for all your help diagnosing this issue!

**Comment by benmcmorran** [comment]
The new setting is available in version 0.6.230214002, which is now available in the pre-release channel in the VS Code Marketplace. From the release notes:

> A new `vcpkg.storageLocation` setting overrides the path to the directory where vcpkg installation files should be stored. The contents of this directory are fully managed by the extension, including potential deletion, so reusing an existing storage location is not recommended. Note that this is not the same as `VCPKG_ROOT`. A `root` directory will be created within this directory to act as the `VCPKG_ROOT`, but additional files will be stored in the directory itself.

To switch to the pre-release channel, simply click the pre-release icon in VS Code.

<img width="276" alt="image" src="https://user-images.githubusercontent.com/6981284/218862102-ffc55b51-44d7-47be-b420-3d1cd7c8f172.png">

**Comment by froha** [comment]
Nice, it is working now.
Thank you!

**Comment by benmcmorran** [comment]
Thanks for confirming! I'm marking this as `fixed-pending-release` since the workaround seems pretty reasonable, but we'll continue to evaluate other options internally to see if there's a better default path we can use for vcpkg storage.

**Comment by benmcmorran** [comment]
The `vcpkg.storageLocation` setting is now available in the official 0.7.0 release.
